### PR TITLE
US5454-Handle Stripe Batches in MP

### PR DESCRIPTION
**Pull request by plachmann**

### CODE REVIEW

File: Gateway/MinistryPlatform.Translation/Repositories/DonationRepository.cs [hunk 1]
```diff
@@ -110,6 +110,14 @@ public MpDonationBatch GetDonationBatchByProcessorTransferId(string processorTra
             
         }
 
+        public MpDeposit GetDepositByProcessorTransferId(string processorTransferId)
+        {
+            var apiToken = _apiUserRepository.GetToken();
+            var searchString = $"Processor_Transfer_ID='{processorTransferId}'";
+
+            return _ministryPlatformRest.UsingAuthenticationToken(apiToken).Search<MpDeposit>(searchString).ToList().FirstOrDefault();
```
Comment (silbermm) [inline Gateway/MinistryPlatform.Translation/Repositories/DonationRepository.cs]: should this guard against null? I.E. `ministryPlatformRest.UsingAuthenticationToken(apiToken).Search<MpDeposit>(searchString).ToList().FirstOrDefault() ?? new MpDeposit()` like is done https://github.com/crdschurch/crds-angular/pull/3188/files#diff-a2338cb5eabe79060366450beb1d9e03R53

File: Gateway/MinistryPlatform.Translation/Repositories/PaymentRepository.cs [hunk 2]
```diff
@@ -43,5 +45,83 @@ public List<MpPayment> GetPaymentsForInvoice(int invoiceId)
...
+            return payment.FirstOrDefault() ?? new MpPayment();
+        }
+
+        public MpPayment GetPaymentById(int paymentId)
+        {
+            var apiToken = _apiUserRepository.GetToken();
+            return _ministryPlatformRest.UsingAuthenticationToken(apiToken).Get<MpPayment>(paymentId);
+        }
+
+        public int UpdatePaymentStatus(int paymentId, int statusId, DateTime dateTime, string statusNote)
```
Comment (silbermm) [inline Gateway/MinistryPlatform.Translation/Repositories/PaymentRepository.cs]: never using `statusNote` of `dateTime`, should we just not even accept them (or are we planning on using in the future)?

File: Gateway/crds-angular/Models/Crossroads/Stewardship/DonationBatchDTO.cs [hunk 3]
```diff
@@ -38,10 +39,19 @@ public class DonationBatchDTO
...
+        public List<PaymentDTO> Payments { get { return (_payments); } }
+        #endregion
+
         [JsonProperty("processor_transfer_id")]
         public string ProcessorTransferId { get; set; }
 
         [JsonProperty("program_id")]
         public int ProgramId { get; set; }
+
+        public decimal BatchFeeTotal { get; set; }
```
Comment (silbermm) [inline Gateway/crds-angular/Models/Crossroads/Stewardship/DonationBatchDTO.cs]: does this need the JsonProperty attribute on it?

File: Gateway/MinistryPlatform.Translation/Repositories/DonationRepository.cs (hunk 1 above)
Comment (plachmann) [inline Gateway/MinistryPlatform.Translation/Repositories/DonationRepository.cs]: Where this function is called it checks for for null and behaves accordingly

File: Gateway/MinistryPlatform.Translation/Repositories/PaymentRepository.cs (hunk 2 above)
Comment (plachmann) [inline Gateway/MinistryPlatform.Translation/Repositories/PaymentRepository.cs]: removed statusNote and datetime

File: Gateway/crds-angular/Models/Crossroads/Stewardship/DonationBatchDTO.cs (hunk 3 above)
Comment (plachmann) [inline Gateway/crds-angular/Models/Crossroads/Stewardship/DonationBatchDTO.cs]: added json property

File: Gateway/crds-angular/Services/StripeEventService.cs [hunk 4]
```diff
@@ -112,9 +114,73 @@ public TransferPaidResponseDTO TransferPaid(DateTime? eventTimestamp, StripeTran
...
+            var donationBatch = CreateBatchDTOFromCharges(donationcharges,depositName+"D",eventTimestamp,transfer, ref response);
+
+            var stripeTotalFees = paymentBatch.BatchFeeTotal + donationBatch.BatchFeeTotal;
+            
+            // Create the deposit
+            var deposit = new DepositDTO
+            {
+                // Account number must be non-null, and non-empty; using a single space to fulfill this requirement
+                AccountNumber = " ",
+                BatchCount = 1,
```
Comment (IngageRitterS) [inline Gateway/crds-angular/Services/StripeEventService.cs]: If we are creating one deposit for potentially two batches, why is the batch count always 1?

File: Gateway/crds-angular.test/Services/StripeEventServiceTest.cs [hunk 5]
```diff
@@ -410,12 +427,14 @@ public void TestTransferPaid()
 
             _donationService.Setup(mocked => mocked.CreateDonationForInvoice(invoice)).Returns(88);
 
-            _donationService.Setup(mocked => mocked.GetDonationBatchByProcessorTransferId("tx9876")).Returns((DonationBatchDTO)null);
-            _paymentService.Setup(mocked => mocked.GetChargesForTransfer("tx9876")).Returns(charges);
+            //_donationService.Setup(mocked => mocked.GetDonationBatchByProcessorTransferId("tx9876")).Returns((DonationBatchDTO)null);
```
Comment (IngageRitterS) [inline Gateway/crds-angular.test/Services/StripeEventServiceTest.cs]: Can this comment be removed?

File: Gateway/crds-angular/Services/StripeEventService.cs [hunk 6]
```diff
@@ -112,9 +114,73 @@ public TransferPaidResponseDTO TransferPaid(DateTime? eventTimestamp, StripeTran
...
+            paymentBatch.DepositId = response.Deposit.Id;
+            donationBatch.DepositId = response.Deposit.Id;
+            try
+            {
+                response.Batch.Add(_paymentService.CreatePaymentBatch(paymentBatch));
+                response.Batch.Add(_donationService.CreateDonationBatch(donationBatch));
+            }
+            catch (Exception e)
+            {
+                _logger.Error("Failed to create donation batch", e);
```
Comment (IngageRitterS) [inline Gateway/crds-angular/Services/StripeEventService.cs]: Message is misleading....was the error in the creation of the donation batch or the payment batch?

File: Gateway/crds-angular/Services/StripeEventService.cs [hunk 7]
```diff
@@ -141,12 +206,12 @@ public TransferPaidResponseDTO TransferPaid(DateTime? eventTimestamp, StripeTran
                     StripeRefund refund = null;
                     if ("refund".Equals(charge.Type)) // Credit Card Refund
                     {
-                        refund = _paymentService.GetChargeRefund(charge.Id);
+                        refund = _paymentProcessorService.GetChargeRefund(charge.Id);
                         paymentId = refund.Data[0].Id;
                     }
                     else if ("payment_refund".Equals(charge.Type)) // Bank Account Refund
```
Comment (IngageRitterS) [inline Gateway/crds-angular/Services/StripeEventService.cs]: Curious what the charge.type will be for a "payment" refund?  Above stripe sends us a type of payment_refund to indicate a bank account refund.

**Pull request merged**